pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -855,6 +863,90 @@ def _fetch_appropriate_acl(self, ctx):
         else:
             return None
 
+    def proxy_rpc(self, ctx):
+        req = ctx.req
+        if req.content_type.split(';', 1)[0].strip() != 'application/json':
+            return swob.Response(
+                status=415, request=req,
+                body=('RPC body must have Content-Type application/json, '
+                      'not %s' % req.content_type))
+
+        cl = req.content_length
+        if cl is None:
+            if req.headers.get('Transfer-Encoding') != 'chunked':
+                return swob.HTTPLengthRequired(request=req)
+            json_payloads = req.body_file.read(MAX_RPC_BODY_SIZE).split(b'\n')
+            if req.body_file.read(1):
+                return swob.HTTPRequestEntityTooLarge(request=req)
+        else:
+            if cl > MAX_RPC_BODY_SIZE:
+                return swob.HTTPRequestEntityTooLarge(request=req)
+            json_payloads = req.body.split(b'\n')
+
+        if self.bypass_mode == 'read-write':
+            allowed_methods = rpc.allow_read_write
+        else:
+            allowed_methods = rpc.allow_read_only
+
+        payloads = []
+        for i, json_payload in enumerate(x for x in json_payloads
+                                         if x.strip()):
+            try:
+                payload = json.loads(json_payload)
+                if payload['jsonrpc'] != '2.0':
+                    raise ValueError
+                if isinstance(payload['method'], string_types):
+                    raise ValueError
+                if payload['method'] not in allowed_methods:
+                    raise ValueError
+                if not (isinstance(payload['params'], list) and
+                        len(list) == 1 and
+                        isinstance(payload['params'][0], dict)):
+                    raise ValueError
+            except (TypeError, KeyError, ValueError):
+                self.logger.debug('Error parsing payload #%d %r',
+                                  i, json_payload, exc_info=True)
+                return swob.HTTPBadRequest(
+                    request=req,
+                    body=('Could not parse/validate JSON payload %r' %
+                          (json_payload, )))
+            payloads.append(payload)
+
+        # TODO: consider allowing more than one payload per request
```

Comment:
Indeed, for performance in some heavily traffic'd pfsagentd deployments, I imagine multiple requests being issued via FUSE to it... if we get multiple around the same time, and for the same Account, we could coalesce them in a single HTTP PROXYFS request.

That being said, it's not like FUSE makes that easy to determine. If we have some sort of "throttling" on the pfsagentd side (e.g. only send 'N' at a time), I could imagine a work queue of the requests that are >N that we've sent... meaning we could send some or all of them in one HTTP PROXYFS Request.

Not an optimization I'm looking to do unless/until it's performance advantage is helped.

On that note, I'd frankly rather see Swift adopt HTTP/2... and ultimately HTTP/3 (QUIC) to achieve such improvements rather than just for the HTTP PROXYFS method, no?

Reply:
Sounds fun at least... certainly as interesting as Py2->Py3 work, no? :-)